aws/request.go

```diff
@@ -217,3 +227,87 @@ func (r *Request) Send() error {
 
 	return nil
 }
+
+func (r *Request) HasNextPage() bool {
+	return r.nextPageToken() != nil
+}
+
+func (r *Request) nextPageToken() interface{} {
+	if r.Operation.Paginator == nil {
+		return nil
+	}
+	v := awsutil.ValuesAtAnyPath(r.Data, r.Operation.OutputToken)
+	if v != nil && len(v) > 0 {
+		return v[0]
+	}
+	return nil
+}
+
+func (r *Request) NextPage() *Request {
+	token := r.nextPageToken()
+	if token == nil {
+		return nil
+	}
+
+	if r.Operation.TruncationToken != "" {
+		tr := awsutil.ValuesAtAnyPath(r.Data, r.Operation.TruncationToken)
+		if tr == nil {
+			return nil
+		} else if len(tr) > 0 {
+			switch v := tr[0].(type) {
+			case bool:
+				if v == false {
+					return nil
+				}
+			}
+		}
+	}
+
+	data := reflect.New(reflect.TypeOf(r.Data).Elem()).Interface()
+	nr := NewRequest(r.Service, r.Operation, awsutil.CopyOf(r.Params), data)
+	awsutil.SetValueAtAnyPath(nr.Params, nr.Operation.InputToken, token)
+	return nr
+}
+
+// This nil value is needed to work around passing nil into reflect.Value#Call()
+// See https://groups.google.com/forum/#!topic/golang-nuts/apNcACpl_fI
+var nilval interface{}
+var rnil = reflect.ValueOf(&nilval).Elem()
+
+func (r *Request) EachPage(fn interface{}) error {
+	valfn := reflect.ValueOf(fn)
+	if valfn.Kind() != reflect.Func {
+		panic("expected function for EachPage()")
```

Comment:
That argument makes sense, and I agree with the differentiation between programmer and runtime errors. But the downside of this form of runtime type checking is that you only catch the code paths you execute in your testing. 99% of the time you have a simple case and you hit the panic in development and fix it right away. But a common issue when accepting an `interface{}` argument is someone passing a pointer when they shouldn't have, or vice versa. And if this happens in an edge case that escaped your testing, you start panicing when releasing to production. Or worst case, code has been live and stable and some external change causes a behavior change which starts executing this bad path.

Of course the dream is 100% test coverage, and recoverer middleware can often stop panics from bubbling to the top for web servers (unless the panic is in a goroutine). But in reality edge cases can go untested, and it's nice when library code is more forgiving.

The Go authors seem to support that stance:
> The convention in the Go libraries is that even when a package uses panic internally, its external API still presents explicit error return values.

The above excerpt is from http://blog.golang.org/defer-panic-and-recover.

All that said, I think it's extremely unlikely anyone would hit these particular panics in production, for the reasons you mentioned. But I'm always a little uncomfortable when I see the possibility of panics escaping libraries.